### PR TITLE
Add ERC: Flash Loans

### PR DESCRIPTION
**Pull request by sambacha**
This PR is just me migrating an existing proposal https://github.com/ethereum/EIPs/blob/d072207e24e3cc12b6315909e6a65275a38e1984/EIPS/eip-7399.md by @alcueca 

The original PR was closed during the ERC/EIP repo split: https://github.com/ethereum/EIPs/pull/7400

Work on this ERC is active, as seen by @@alcueca repo here: https://github.com/alcueca/erc7399-wrappers

How many more of these instances are there possibly?

### CODE REVIEW

**Comment by alcueca** [comment]
If possible, we would like to have it merged as draft. I think that all the work was already done for that.

There is an outstanding issue with Vyper not implementing function arguments, so this ERC might stay as draft while that gets resolved.

File: ERCS/erc-7399.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+ - Syntax of the call initiating the flash loan.
+ - Possibility of the loan receiver and the callback receiver being different addresses.
+ - Repayment approaches at the end of the transaction, where some pull the principal plus the fee from the loan receiver, and others where the loan receiver needs to manually return the principal and the fee to the lender.
+ - Some lenders allow to flash mint any amount of their native asset without charging a fee, effectively allowing flash loans bounded by computational constraints instead of asset ownership constraints.
+
+This ERC aims to consolidate all flash lenders into a common standard of maximum flexibility, so that borrowers can easily choose flash lenders without changing their code.
+
+## Specification
+
```
Comment (gaudren) [inline ERCS/erc-7399.md:32]:
```suggestion

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

```

File: ERCS/erc-7399.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+ The `amount` and `fee` need to be transferred to a `repayment receiver` before the end of the `flash` call. The `fee` can be set to zero `asset`.
+
+The `callback function` can return any arbitrary data which will be received by the `initiator` as the return value of the `flash` call.
+
+### Lender Specification
+
+A `lender` MUST implement the [ERC-7399](./eip-7399.md) interface.
+
+```solidity
+// SPDX-License-Identifier: CC0
```
Comment (gaudren) [inline ERCS/erc-7399.md:52]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: ERCS/erc-7399.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+A `fee` will often be calculated in the callback function, which the callback receiver must be aware of for repayment. Passing the `fee` as a parameter instead of appended to `data` is simple and effective.
+
+Arbitrary callback functions on callback receivers allows to implement different behaviours to flash loans on callback receivers without the need for encoding a function router using the `data` argument. A function call type is 24 bytes of which the first 20 bytes are the target address and the last 4 bytes are the function signature.
+
+The `amount + fee` are pushed to the `payment receiver` to allow for the segregation of asset and control flows. While a "pull" architecture is more prevalent, "push" architectures are also common. For those cases where the `lender` can't implement a "push" architecture, a simple wrapper contract can offer an ERC-7399 external interface, while using liquidity from the `lender` using a "pull" architecture.
```
Comment (gaudren) [inline ERCS/erc-7399.md:167]:
```suggestion
The `amount + fee` are pushed to the `payment receiver` to allow for the segregation of asset and control flows. While a "pull" architecture is more prevalent, "push" architectures are also common. For those cases where the `lender` can't implement a "push" architecture, a simple wrapper contract can offer this proposal's external interface, while using liquidity from the `lender` using a "pull" architecture.
```

File: ERCS/erc-7399.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+## Backwards Compatibility
+
+This EIP is a successor of [ERC-3156](./eip-3156.md). While not directly backwards compatible, a wrapper contract offering a ERC-7399 external interface with liquidity obtained from an ERC-3156 flash `lender` is trivial to implement.
```
Comment (gaudren) [inline ERCS/erc-7399.md:171]:
```suggestion
This EIP is a successor of [ERC-3156](./eip-3156.md). While not directly backwards compatible, a wrapper contract offering this proposal's external interface with liquidity obtained from an ERC-3156 flash `lender` is trivial to implement.
```

**Comment by abcoathup** [comment]
@sambacha

> This PR is just me migrating an existing proposal https://github.com/ethereum/EIPs/blob/d072207e24e3cc12b6315909e6a65275a38e1984/EIPS/eip-7399.md by @alcueca

You should get approval from the original authors (assuming you aren't listed as one of them) to include you as an author, given changes are probably required prior to merging the PR.  cc: @alcueca, @ultrasecreth, @devtooligan, @AmadiMichaels

**Comment by alcueca** [comment]
Hey @sambacha, would you like to add yourself as a co-author, please?

Feel free to also solve the other comments above, to the best of your ability, I'd appreciate it.

**Comment by sambacha** [comment]
> Hey @sambacha, would you like to add yourself as a co-author, please?
> 
> Feel free to also solve the other comments above, to the best of your ability, I'd appreciate it.

https://github.com/alcueca/erc7399-wrappers

Ill sort this out this weekend!

happy easter 🐣

File: ERCS/erc-7399.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
+---
+eip: 7399
+title: Flash Loans
```
[the remark this answers is not in the document]
Comment (alcueca) [reply ERCS/erc-7399.md:3]: @sambacha: Flexible Flash Loans?

Basically the difference between ERC3156 and ERC3799 is that the latter is a lot more flexible (giving power to users in exchange of some complexity).

File: ERCS/erc-7399.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+description: Interfaces and processes for flash loans
```
[the remark this answers is not in the document]
Comment (alcueca) [reply ERCS/erc-7399.md:4]: @sambacha, maybe we just copy the abstract from erc7399?

```
A flash loan is a smart contract transaction in which a lender smart contract lends assets to a borrower smart contract with the condition that the assets are returned, plus an optional fee, before the end of the transaction. This ERC specifies interfaces for lenders to accept flash loan requests, and for borrowers to take temporary control of the transaction within the lender execution. The process for the safe execution of flash loans is also specified.
```

File: ERCS/erc-7399.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+## Motivation
+
+Early adopters of the flash loan pattern have produced different interfaces and different use patterns. The diversification is expected to intensify, and with it the technical debt required to integrate with diverse flash lending patterns.
+
+Some of the high level differences in the approaches across the protocols include:
+
+ - Syntax of the call initiating the flash loan.
+ - Possibility of the loan receiver and the callback receiver being different addresses.
+ - Repayment approaches at the end of the transaction, where some pull the principal plus the fee from the loan receiver, and others where the loan receiver needs to manually return the principal and the fee to the lender.
```
[the remark this answers is not in the document]
Comment (alcueca) [reply ERCS/erc-7399.md:26]: @sambacha, given that we introduced this bullet point list as "Some of the high level differences in the approaches across the protocols include:"

Maybe we can say something like:

"Flash loan repayment mechanism, which can be pull (take the repayment from the borrower) or push (expect the borrower to send the repayment to the lender)".

**Comment by sambacha** [comment]
@gaudren can you re-review please sir?

**Comment by sambacha** [comment]
> Hey @sambacha, would you like to add yourself as a co-author, please?
> 
> Feel free to also solve the other comments above, to the best of your ability, I'd appreciate it.

The only outstanding issue left is the title for this proposal, ser.

**Review by sambacha — COMMENTED** [review]
@alcueca what say you ser!

File: ERCS/erc-7399.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaudren) [reply ERCS/erc-7399.md:4]: Your description should explicitly say what makes your proposal different from existing flash loans. I'm not familiar enough with DeFi to give a useful suggestion.

**Review by gaudren — APPROVED** [review]
This looks good to me. You can work on your description while in draft.

**Review by SamWilsn — APPROVED** [review]
Looks good for a draft. Like @gaudren said, please try to add enough information to your title and description to differentiate this proposal from other similar ones before moving to review.

**Comment by sambacha** [comment]
> This looks good to me. You can work on your description while in draft.

What if we add an emoji, like "⚡️ Flash Loan"
Come to think of it I do not know of any EIPs with an emoji in their title, that should be sufficiently different!

**Pull request merged**